oci/oci.go

```diff
@@ -235,6 +219,254 @@ func (r *RuntimeBase) path(c *Container) (string, error) {
+		close(chControl)
+		return fmt.Errorf("failed to get container stopped status: %ds timeout reached", timeout)
+	}
+
+	if err != nil {
+		return fmt.Errorf("failed to get container stopped status: %v", err)
+	}
+
+	return nil
+}
+
+func (r *Runtime) newRuntimeImpl(c *Container) (runtimeImpl, error) {
+	rPath, err := r.path(c)
+	if err != nil {
+		return nil, err
+	}
+
+	if c.runtimeHandler != "" {
+		runtimeHandler, err := r.ValidateRuntimeHandler(c.runtimeHandler)
+		if err == nil && runtimeHandler.RuntimeType == RuntimeTypeVM {
+			return newRuntimeVM(rPath), nil
+		}
+	}
+
+	if r.runtimeType == RuntimeTypeVM {
+		return newRuntimeVM(rPath), nil
+	}
+
+	return newRuntimeOCI(r, rPath), nil
+}
+
+func (r *Runtime) runtimeImpl(c *Container) (runtimeImpl, error) {
+	impl, ok := r.runtimeImplList[c.ID()]
+	if !ok {
+		return r.newRuntimeImpl(c)
+	}
+
+	return impl, nil
+}
+
+// CreateContainer creates a container.
+func (r *Runtime) CreateContainer(c *Container, cgroupParent string) error {
+	// Instantiate a new runtime implementation for this new container
+	impl, err := r.newRuntimeImpl(c)
+	if err != nil {
+		return err
+	}
+
+	// Assign this runtime implementation to the current container
+	r.runtimeImplList[c.ID()] = impl
+
+	return impl.createContainer(c, cgroupParent)
+}
+
+// StartContainer starts a container.
+func (r *Runtime) StartContainer(c *Container) error {
+	impl, err := r.runtimeImpl(c)
+	if err != nil {
+		return err
+	}
+
+	return impl.startContainer(c)
+}
+
+// ExecContainer prepares a streaming endpoint to execute a command in the container.
+func (r *Runtime) ExecContainer(c *Container, cmd []string, stdin io.Reader, stdout, stderr io.WriteCloser, tty bool, resize <-chan remotecommand.TerminalSize) error {
+	impl, err := r.runtimeImpl(c)
+	if err != nil {
+		return err
+	}
+
+	return impl.execContainer(c, cmd, stdin, stdout, stderr, tty, resize)
+}
+
+// ExecSyncContainer execs a command in a container and returns it's stdout, stderr and return code.
+func (r *Runtime) ExecSyncContainer(c *Container, command []string, timeout int64) (*ExecSyncResponse, error) {
+	impl, err := r.runtimeImpl(c)
+	if err != nil {
+		return nil, err
+	}
+
+	return impl.execSyncContainer(c, command, timeout)
+}
+
+// UpdateContainer updates container resources
+func (r *Runtime) UpdateContainer(c *Container, res *rspec.LinuxResources) error {
+	impl, err := r.runtimeImpl(c)
+	if err != nil {
+		return err
+	}
+
+	return impl.updateContainer(c, res)
+}
+
+// StopContainer stops a container. Timeout is given in seconds.
+func (r *Runtime) StopContainer(ctx context.Context, c *Container, timeout int64) error {
+	impl, err := r.runtimeImpl(c)
+	if err != nil {
+		return err
+	}
+
+	return impl.stopContainer(ctx, c, timeout)
+}
+
+// DeleteContainer deletes a container.
+func (r *Runtime) DeleteContainer(c *Container) error {
+	impl, err := r.runtimeImpl(c)
+	if err != nil {
+		return err
+	}
+
+	defer delete(r.runtimeImplList, c.ID())
```

Comment:
Do we want to delete this impl if deleteContainer fails?

Reply:
That's a good point. I can make this defer function behaving like that instead:
```go
func (r *Runtime) DeleteContainer(c *Container) (err error) {
	impl, err := r.runtimeImpl(c)
	if err != nil {
		return err
	}

	defer func() {
                if err == nil {
                        delete(r.runtimeImplList, c.ID())
                }
        }()

	return impl.deleteContainer(c)
}
```

But also I'm not sure. This would mean a failure would result in the implementation map to hold onto stale values.
@mrunalp wdyt?

Reply:
I thikn you'll have to assign
`err = impl.deleteContainer(c)
return err` 
for that solution, but yeah that should work (if deleting the runtime is  something to avoid in that case, I'm not particularly sure

Reply:
though I think at that point it'd be cleaner to do
```
func (r *Runtime) DeleteContainer(c *Container) (err error) {
	impl, err := r.runtimeImpl(c)
	if err != nil {
		return err
	}
        err = impl.deleteContainer(c)
        if err == nil {
                delete(r.runtimeImplList, c.ID())
        }
        return err
}

Reply:
Are you sure I would have to do so. I thought the `return` assumes the `err` get assigned with the value returned by `impl.deleteContainer(c)`. And I think this behavior is implied by the declaration of `(err error)` as part of the prototype.
But I'm not entirely sure, and if we decide to implement it this way, I'll double check!

Reply:
Yeah true, no need to use defer here :)

Reply:
oh yeah I missed the (err error) and was basing it off of just returning a generic error, you're right in that case haha

Reply:
@mrunalp wdyt about the behavior that we should follow here?

Reply:
I think we should remove it only if delete doesn't return any errors.